File: pdf-php/examples/generate_images.php
<?php
/**
 * Image support example — JPEG, PNG, and RGBA PNG with four fit modes.
 *
 * Mirrors: pdf-core/examples/generate_images.rs
 *
 * Run with:
 *   php -d extension=target/release/libpdf_php.so pdf-php/examples/generate_images.php
 */

@mkdir(__DIR__ . '/../../output', 0755, true);
$path        = __DIR__ . '/../../output/php-images.pdf';
$fixturesDir = __DIR__ . '/../../pdf-core/tests/fixtures';

$doc = PdfDocument::create($path);
$doc->setCompression(true);
$doc->setInfo("Creator", "rust-pdf-php");
$doc->setInfo("Title", "Image Support Demo");

// Load images
$jpeg     = $doc->loadImageFile("$fixturesDir/test.jpg");
$png      = $doc->loadImageFile("$fixturesDir/test.png");
$pngAlpha = $doc->loadImageFile("$fixturesDir/test_alpha.png");

// --- Page 1: all four fit modes ---
$doc->beginPage(612.0, 792.0);
$doc->placeText("Image Support Demo", 72.0, 750.0);

// Fit mode — scales to fit, preserves aspect ratio
$doc->placeText("Fit (JPEG)", 72.0, 700.0);
$doc->placeImage($jpeg, new Rect(72.0, 100.0, 200.0, 150.0), "fit");

// Stretch mode — fills rect exactly, may distort
$doc->placeText("Stretch (PNG)", 320.0, 700.0);
$doc->placeImage($png, new Rect(320.0, 100.0, 200.0, 150.0), "stretch");

// Fill mode — scales to cover, clips overflow
$doc->placeText("Fill (PNG Alpha)", 72.0, 480.0);
$doc->placeImage($pngAlpha, new Rect(72.0, 320.0, 200.0, 150.0), "fill");

// None mode — natural size (1px = 1pt)
$doc->placeText("None (PNG)", 320.0, 480.0);
$doc->placeImage($png, new Rect(320.0, 320.0, 200.0, 150.0), "none");

$doc->endPage();

// --- Page 2: same JPEG (demonstrates XObject reuse) ---
$doc->beginPage(612.0, 792.0);
$doc->placeText("Same JPEG on page 2 (XObject reused)", 72.0, 750.0);
$doc->placeImage($jpeg, new Rect(72.0, 100.0, 468.0, 600.0), "fit");
$doc->endPage();

$doc->endDocument();

echo "Generated: $path\n";

File: examples/rust/generate_images.rs
use pdf_core::{ImageFit, PdfDocument, Rect};

fn main() {
    std::fs::create_dir_all("examples/output").unwrap();
    let path = "examples/output/rust-images.pdf";
    let mut doc = PdfDocument::create(path).unwrap();
    doc.set_compression(true);
    doc.set_info("Creator", "rust-pdf");
    doc.set_info("Title", "Image Support Demo");

    // Load images
    let jpeg = doc
        .load_image_file("pdf-core/tests/fixtures/test.jpg")
        .unwrap();
    let png = doc
        .load_image_file("pdf-core/tests/fixtures/test.png")
        .unwrap();
    let png_alpha = doc
        .load_image_file("pdf-core/tests/fixtures/test_alpha.png")
        .unwrap();

    doc.begin_page(612.0, 792.0);
    doc.place_text("Image Support Demo", 72.0, 750.0);

    // Fit mode — scales to fit, preserves aspect ratio
    doc.place_text("Fit (JPEG)", 72.0, 700.0);
    let r1 = Rect {
        x: 72.0,
        y: 100.0,
        width: 200.0,
        height: 150.0,
    };
    doc.place_image(&jpeg, &r1, ImageFit::Fit);

    // Stretch mode — fills rect exactly, may distort
    doc.place_text("Stretch (PNG)", 320.0, 700.0);
    let r2 = Rect {
        x: 320.0,
        y: 100.0,
        width: 200.0,
        height: 150.0,
    };
    doc.place_image(&png, &r2, ImageFit::Stretch);

    // Fill mode — scales to cover, clips overflow
    doc.place_text("Fill (PNG Alpha)", 72.0, 480.0);
    let r3 = Rect {
        x: 72.0,
        y: 320.0,
        width: 200.0,
        height: 150.0,
    };
    doc.place_image(&png_alpha, &r3, ImageFit::Fill);

    // None mode — natural size (1px = 1pt)
    doc.place_text("None (PNG)", 320.0, 480.0);
    let r4 = Rect {
        x: 320.0,
        y: 320.0,
        width: 200.0,
        height: 150.0,
    };
    doc.place_image(&png, &r4, ImageFit::None);

    // Same image on a second page (demonstrates write-once)
    doc.begin_page(612.0, 792.0);
    doc.place_text("Same JPEG on page 2 (XObject reused)", 72.0, 750.0);
    let r5 = Rect {
        x: 72.0,
        y: 100.0,
        width: 468.0,
        height: 600.0,
    };
    doc.place_image(&jpeg, &r5, ImageFit::Fit);

    doc.end_page().unwrap();
    doc.end_document().unwrap();
    println!("Generated: {}", path);
}
